src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,432 @@
+
+handle_cast(
+    write,
+    State = #state{
+        write_dir = WriteDir,
+        file_path = FilePath,
+        max_file_size = MaxFileSize,
+        upload_window_start_time = WindowStartTime,
+        upload_window = UploadWindow
+    }
+) ->
+    {ok, FileSize} = handle_write(FilePath, MaxFileSize, ?FILE_WRITE_OPTIONS),
+
+    case FileSize >= MaxFileSize orelse upload_window_elapsed(WindowStartTime, UploadWindow) of
+        true ->
+            ?MODULE:upload_packets(FilePath),
+            {noreply, State#state{file_path = generate_file_name(WriteDir)}};
+        false ->
+            {noreply, State}
+    end;
+handle_cast({upload, FilePath}, State = #state{aws_client = AWSClient, bucket = Bucket}) ->
+    UploadTimestamp = erlang:system_time(millisecond),
+    FileName = list_to_binary("packetreport." ++ integer_to_list(UploadTimestamp) ++ ".gz"),
+
+    case upload_file(AWSClient, list_to_binary(FilePath), FileName, Bucket) of
+        {ok, _} ->
+            file:delete(FilePath);
+        Error ->
+            lager:warning("packet reporter failed to upload: ~p~n", [Error]),
+            error
+    end,
+    {noreply, State#state{upload_window_start_time = UploadTimestamp}};
+handle_cast({start_interval, IntervalDuration}, State = #state{report_interval = ReportInterval}) ->
+    {ok, ReportInterval2} = handle_restart_interval(ReportInterval, IntervalDuration),
+    {noreply, State#state{report_interval = ReportInterval2, interval_duration = IntervalDuration}};
+handle_cast(stop_interval, State = #state{report_interval = ReportInterval}) ->
+    handle_stop_interval(ReportInterval),
+    {noreply, State#state{report_interval = undefined, interval_duration = undefined}};
+handle_cast(_Msg, State = #state{}) ->
+    {noreply, State}.
+
+handle_info(_Info, State = #state{}) ->
+    {noreply, State}.
+
+terminate(
+    Reason,
+    #state{
+        file_path = FilePath,
+        report_interval = ReportInterval,
+        max_file_size = MaxFileSize
+    }
+) ->
+    handle_write(FilePath, MaxFileSize, ?FILE_WRITE_OPTIONS),
+    handle_stop_interval(ReportInterval),
+    lager:warning("packet reporter process terminated: ~s", [Reason]),
+    ok.
+
+code_change(_OldVsn, State = #state{}, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal Function Definitions
+%%%===================================================================
+
+-spec handle_write(FilePath :: string(), MaxFileSize :: integer(), WriteOptions :: [atom()]) ->
+    {ok, FileSize :: integer()}.
+
+handle_write(FilePath, MaxFileSize, WriteOptions) ->
+    WriteTimestamp = erlang:system_time(millisecond),
+    {ok, S} = open_tmp_file(FilePath, WriteOptions),
+    Data = get_packets_by_timestamp(WriteTimestamp),
+
+    lists:foreach(
+        fun(Packet) ->
+            PacketSize = encode_packet_size(Packet),
+            file:write(S, PacketSize),
+            file:write(S, Packet)
```

Comment:
I think you can combine these into a single call: `file:write(S, [PacketSize, Packet])`. The data for `file:write` is `iodata()`.